### ISSUE TITLE
Assertion failure (pattern matching) 4.12 & trunk

### ISSUE THREAD

**Issue by nojb** [ocaml/ocaml#10294]
cc @trefis @gasche 
```
PERVERSESHEAF$ cat foo.ml
type import_error =
  | Node of string

type export_error =
  | Variant of string * string

exception Excel_import of import_error
exception Excel_export of export_error

let f = function
  | Excel_import (Node _) ->
      None
  | Excel_export (Variant _) ->
      None
  | _ ->
      None

PERVERSESHEAF$ ocamlc -c foo.ml
Fatal error: exception File "lambda/matching.ml", line 705, characters 18-24: Assertion failed
```

**Comment by nojb** [comment]
Replacing the pattern variables by wildcards `_` makes the failure go away.

**Comment by gasche** [comment]
Thanks for the report! I proposed a preliminary fix in #10295.

**Comment by maranget** [comment]
I am wondering, would it be possible to

**Comment by nojb** [comment]
(assuming it was closed by mistake)

**Comment by maranget** [comment]
Right, I hit the close button, aiming at the cancel one.  Thanks for fixing this.

**Comment by Octachron** [comment]
This issue was fixed by merging #10295 .
